app/lib/search/updater.dart

```diff
@@ -4,92 +4,107 @@
 
 import 'dart:async';
 
-import 'package:gcloud/db.dart';
 import 'package:logging/logging.dart';
 
-import '../frontend/models.dart';
-import '../shared/mock_scores.dart';
 import '../shared/search_service.dart';
-import '../shared/utils.dart' show sliceList;
+import '../shared/task_scheduler.dart';
+
+import 'backend.dart';
+import 'index_ducene.dart';
 
 Logger _logger = new Logger('pub.search.updater');
 
-class PackageIndexUpdater {
-  final PackageIndex index;
-  Timer _timer;
-  bool _updating = false;
+// A shared state between [IndexUpdateTaskSource] and [BatchIndexUpdater],
+// indicating how many packages were yielded in the first run of the index
+// update.
+// When [BatchIndexUpdater] process more than this number of tasks, it will
+// start do the index merges, making sure that the index is marked as ready.
+int _firstPackageListCount;
 
-  PackageIndexUpdater(this.index);
+class IndexUpdateTaskSource implements TaskSource {
+  final bool isLocalDevMode;
+  DateTime _lastTs;
+  IndexUpdateTaskSource({this.isLocalDevMode: false});
 
-  Future update({int limit}) async {
-    if (_updating) {
-      _logger.warning('Update called before previous round completed.');
-      return;
-    }
-    try {
-      _updating = true;
-      await _doUpdate(limit);
-    } catch (e, st) {
-      _logger.severe('Error while updating search index.', e, st);
-    } finally {
-      _updating = false;
+  @override
+  Stream<Task> startStreaming() async* {
+    for (;;) {
+      final DateTime now = new DateTime.now().toUtc();
+      int counter = 0;
+      if (isLocalDevMode && counter > 300) break;
+      await for (String package
+          in searchBackend.listPackages(updatedAfter: _lastTs)) {
+        counter++;
+        yield new Task(package, null);
+        if (isLocalDevMode && counter > 300) break;
```

Comment:
I was using this limit to make the fine-tuning of the indexing and search snappier. Re-indexing the entire staging each time I made a small change was taking long, and this was a reasonable limit. There are other ways for that, so I'll just remove this part of the logic.